### ISSUE TITLE
Weights and Biases support

### ISSUE THREAD

**Issue by Guitaricet** [pytorch/fairseq#1790]
## 🚀 Feature Request
Add [WandB](https://www.wandb.com/) support for logging.

### Motivation

- WandB is cloud-based - you can access your logs without a direct ssh connection, just using the website
- WandB interface is better suited for many experiments, it also can log model weights, configs, and source code
- WandB is prettier

### Pitch

Tensorboard is slow a buggy - sometimes I wait about a minute for it to load all the logs, sometimes it just doesn't show them.
There is no reason not to allow fairseq users to use wandb.
It is also very simple to add a new progress bar wrapper (and I already did that).

### Alternatives

Logger extension interface (plug-in) which can allow users to use any logging method.

### Additional context

I already coded this feature for myself. Currently, I am using it for my experiments and it looks pretty good. You can see PR draft  #1781 for more information.
I'll be happy if you can help me to understand why Windows build is failing, I did not write any platform-specific code and I have no windows computer to test it.

**Comment by Guitaricet** [comment]
Anybody? =(

**Comment by CSP197** [comment]
@Guitaricet - I can help with this and I have a Windows computer for testing =)

**Comment by Guitaricet** [comment]
Awesome! You can look into [this](https://github.com/pytorch/fairseq/runs/490946509) ci log. Tests are failing for `windows-latest, 3.6`. Everything is looking good for Ubuntu and MacOS. I also double-checked MacOS tests locally on my machine.

Also, what kind of tests does wandb integration require (if any)? I did not find any tests for tensorboard.

**Comment by Guitaricet** [comment]
@CSP197 thank you for offering help, but I just noticed that this test is failing on all PRs, including documentation improvements. So this is not a problem of this one.

**Comment by CSP197** [comment]
@Guitaricet - Yeah, I think its a permissions issue with the Windows system.
I can run the tests successfully on my local machine...

` Ran 116 tests in 166.650s
OK (skipped=17)  `

...but run into the same issues when running on the Github Actions workflow.

**Comment by Guitaricet** [comment]
@CSP197 thank you!
Now I hope somebody from fairseq will find this feature useful.

**Comment by Guitaricet** [comment]
PR is ready for review

**Comment by Guitaricet** [comment]
@myleott sorry that I am asking you directly, but could you please comment if you want this feature in fairseq and accept the PR if so.

**Comment by BadrYoubiIdrissi** [comment]
Would love to see this feature added :) Although maybe exposing a logging extension interface might be more appropriate. It would avoid core fairseq from having many cutom logging progress bars.

**Comment by Guitaricet** [comment]
@BadrYoubiIdrissi I improved my implementation a bit - now it also logs the model architecture, gradient norms, and the logger output before the training. The fork is also updated. Hope you'll find it useful.

**Comment by rmrao** [comment]
Would be really nice to see this merged!

**Comment by myleott** [comment]
Merged in 1a709b2a401ac8bd6d805c8a6a5f4d7f03b923ff

**Comment by zwhe99** [comment]
Can wandb be used in distributed training?
